### PR TITLE
add imagePullSecrets to sidecar injection template

### PR DESCRIPTION
**Pull request by tariq1890**
Since `imagePullSecrets` is a part of the `SidecarInjectionSpec`, it would be good to illustrate the usage of this through the manifest charts. Moreover, global values are used as settings in the sidecar injection template, so it only makes sense that we use the same pattern for providing the `imagePullSecrets` via `.Values.global`

[X] Configuration Infrastructure
[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Review by morvencao — COMMENTED** [review]
Have you tested this case? What if the original pod already has `imagePullSecrets`?

**Review by howardjohn — COMMENTED** [review]
won't you need to add this secret to every namespace?

We should at least add a test under inject_test.go for this if we are going to add this.

**Comment by tariq1890** [comment]
> Have you tested this case? What if the original pod already has imagePullSecrets?

In this case, it should append to the list of imagePullSecrets. I thought the code already does that?

**Comment by tariq1890** [comment]
> won't you need to add this secret to every namespace?

I think this is OK to assume. If any istio participating workload needs to pull down the istio-proxy image, the pullSecrets will have to be made available in all namespaces

**Comment by howardjohn** [comment]
> In this case, it should append to the list of imagePullSecrets. I thought the code already does that?

If you add a test then we won't have to guess :slightly_smiling_face:

File: pkg/kube/inject/inject.go [hunk 1]
```diff
@@ -761,6 +761,8 @@ func IntoObject(sidecarTemplate string, valuesConfig string, revision string, me
 
 	podSpec.DNSConfig = spec.DNSConfig
 
+	podSpec.ImagePullSecrets = append(podSpec.ImagePullSecrets, spec.ImagePullSecrets...)
```
Comment (tariq1890) [inline pkg/kube/inject/inject.go:764]: Thanks to the unit tests, I added this bugfix :).

Good call @howardjohn !

**Review by howardjohn — APPROVED** [review]

**Comment by tariq1890** [comment]
Thanks for the approve @howardjohn . Do you agree this needs to be cherrypicked to release-1.6?

**Review by sdake — APPROVED** [review]
LGTM. At some point, @istio/wg-environments-maintainers needs to have a discussion and agreement about "one API" to rule hem all. Currently we have three, and this PR adds to the deprecated API `global.yaml`. I understand completely why this option is necessary, although at some point, we need to consider a reorg of all of the 3 APIs (values,yaml, IsitoOperator, and MeshConfig) into one API.

The current (fuzzy) agreement as I understand it is to push everything to `MeshConfig`. I am not convinced this is a good choice as `MeshConfig` is already poorly organized.

What I'd like to see is one new object, that every component understands, that meets the criteria for a stable API. This viewpoint means deprecating `MeshConfig` in favor of `IstioConfig` - which would be a new take on IstioOperator - but with all of the stuff in `MeshConfig`.

For now, we can make an exception and add to values.yaml - but again, the WG needs to have further discussion and a concrete agreement on one API - used throughout istiod (mostly PIlot), as well as istioctl and the standalone operator and the helm charts.

Cheers,
-steve

**Comment by sdake** [comment]
cc / @rcernich I know you may be doing some work in this area. Also @ostromart is looking into this deeply.

I can bring us together - but would need an initial proposal.

**Comment by howardjohn** [comment]
`.global.imagePullSecrets` is not a new option, it has existed forever. Just not applied to injection template

**Comment by sdake** [comment]
@tariq1890 I've selected this PR for backport to 1.6.

Cheers,
-steve

**Comment by tariq1890** [comment]
https://github.com/istio/istio/pull/24446

**Comment by tariq1890** [comment]
Thanks @sdake @howardjohn

**Comment by rcernich** [comment]
@sdake, I'm working on a POC that reorganizes the settings in values.yaml based on feature/area of concern.  I've also drafted an RFC, which should be ready for review next week.

**Pull request merged**